cmd/registry/conformance/conformance-task.go

```diff
@@ -0,0 +1,210 @@
+
+	// Download spec
+
+	data, err := core.GetBytesForSpec(ctx, task.Client, task.Spec)
+	if err != nil {
+		return err
+	}
+	// Put the spec in a temporary directory.
+	root, err := ioutil.TempDir("", "registry-spec-")
+	if err != nil {
+		return err
+	}
+	name := filepath.Base(task.Spec.GetName())
+	defer os.RemoveAll(root)
+
+	if core.IsZipArchive(task.Spec.GetMimeType()) {
+		_, err = core.UnzipArchiveToPath(data, root)
+	} else {
+		// Write the file to the temporary directory.
+		err = ioutil.WriteFile(filepath.Join(root, name), data, 0644)
+	}
+	if err != nil {
+		return err
+	}
+
+	// Generate conformance report
+	conformanceReport := initializeConformanceReport(task.Spec.GetName(), task.StyleguideId)
+	err = task.computeConformanceReport(ctx, root, conformanceReport)
+	if err != nil {
+		log.Errorf(ctx,
+			"Failed to compute the conformance for spec %s: %s",
+			task.Spec.GetName(),
+			err,
+		)
+		return err
+	}
+
+	return task.storeConformanceReport(ctx, conformanceReport)
+}
+
+func (task *ComputeConformanceTask) computeConformanceReport(
+	ctx context.Context,
+	specDirectory string,
+	conformanceReport *rpc.ConformanceReport,
+) error {
+
+	guidelineReportsMap := make(map[string]*rpc.GuidelineReport)
+
+	// Iterate over all the linters, and lint.
+	for _, metadata := range task.LintersMetadata {
+
+		linterResponse, err := invokeLinter(ctx, specDirectory, metadata)
+		// If a linter returned an error, we shouldn't stop linting completely across all linters and
+		// discard the conformance report for this spec. We should log but still continue, because there
+		// may still be useful information from other linters that we may be discarding.
+		if err != nil {
+			log.Errorf(ctx, "Linter error: %s", err)
+			continue
+		}
+
+		// Process linterResponse to generate conformance report
+		lintFiles := linterResponse.Lint.GetFiles()
+
+		for _, lintFile := range lintFiles {
+			lintProblems := lintFile.GetProblems()
+
+			// Iterate over the list of problems returned by the linter.
+			for _, problem := range lintProblems {
+				ruleMetadata, ok := metadata.rulesMetadata[problem.GetRuleId()]
+				if !ok {
+					// If a problem the linter returned isn't one that we expect
+					// then we should ignore it
+					continue
+				}
+
+				guideline := ruleMetadata.guideline
+				guidelineRule := ruleMetadata.guidelineRule
+
+				// Check if the guideline report for the guideline which contains this rule
+				// has already been initialized. If it hasn't then create one.
+				guidelineReport, ok := guidelineReportsMap[guideline.GetId()]
+				if !ok {
+					guidelineReport = initializeGuidelineReport(guideline.GetId())
+
+					guidelineReportsMap[guideline.GetId()] = guidelineReport
+					conformanceReport.GuidelineReportGroups[guideline.Status].GuidelineReports =
+						append(conformanceReport.GuidelineReportGroups[guideline.Status].GuidelineReports, guidelineReport)
```

Comment:
Can this be split into smaller steps? There's a lot going on in this one statement. Maybe assigning the group to a variable before appending to its GuidelineReports slice would simplify it a good amount.

Reply:
Done